lib/libstatcounters/statcounters.c

```diff
@@ -28,643 +28,256 @@
-// instruction cache, CacheCore, 8
-// data cache, CacheCore, 9
-// level 2 cache, CacheCore, 10
-// memory operations, MIPSMem, 11
-// tag cache, CacheCore, 12
-// level 2 cache master, MasterStats, 13
-// tag cache master, MasterStats, 14
-
-// module type : CacheCore
-// available counter, associated rdhwr secondary selector
-//----------------------------------------------------------------------------
-// write hit, 0
-// write miss, 1
-// read hit, 2
-// read miss, 3
-// prefetch hit, 4
-// prefetch miss, 5
-// eviction, 6
-// eviction due to prefetch, 7
-// write of tag set, 8
-// read of tag set, 9
-enum
-{
-    STATCOUNTERS_WRITE_HIT     = 0,
-    STATCOUNTERS_WRITE_MISS    = 1,
-    STATCOUNTERS_READ_HIT      = 2,
-    STATCOUNTERS_READ_MISS     = 3,
-    STATCOUNTERS_PFTCH_HIT     = 4,
-    STATCOUNTERS_PFTCH_MISS    = 5,
-    STATCOUNTERS_EVICT         = 6,
-    STATCOUNTERS_PFTCH_EVICT   = 7,
-    STATCOUNTERS_SET_TAG_WRITE = 8,
-    STATCOUNTERS_SET_TAG_READ  = 9
-};
-//----------------------------------------------------------------------------
-// module type : MIPSMem
-// available counter, associated rdhwr secondary selector
-//----------------------------------------------------------------------------
-// byte read, 0
-// byte write, 1
-// half word read, 2
-// half word write, 3
-// word read, 4
-// word write, 5
-// double word read, 6
-// double word write, 7
-// capability read, 8
-// capability write, 9
-enum
-{
-    STATCOUNTERS_BYTE_READ         = 0,
-    STATCOUNTERS_BYTE_WRITE        = 1,
-    STATCOUNTERS_HWORD_READ        = 2,
-    STATCOUNTERS_HWORD_WRITE       = 3,
-    STATCOUNTERS_WORD_READ         = 4,
-    STATCOUNTERS_WORD_WRITE        = 5,
-    STATCOUNTERS_DWORD_READ        = 6,
-    STATCOUNTERS_DWORD_WRITE       = 7,
-    STATCOUNTERS_CAP_READ          = 8,
-    STATCOUNTERS_CAP_WRITE         = 9,
-    STATCOUNTERS_CAP_READ_TAG_SET  = 10,
-    STATCOUNTERS_CAP_WRITE_TAG_SET = 11
-};
-//----------------------------------------------------------------------------
-// module type : MasterStats
-// available counter, associated rdhwr secondary selector
-//----------------------------------------------------------------------------
-// read request, 0
-// write request, 1
-// write request flit, 2
-// read response, 3
-// read response flit, 4
-// write response, 5
-enum
-{
-    STATCOUNTERS_READ_REQ       = 0,
-    STATCOUNTERS_WRITE_REQ      = 1,
-    STATCOUNTERS_WRITE_REQ_FLIT = 2,
-    STATCOUNTERS_READ_RSP       = 3,
-    STATCOUNTERS_READ_RSP_FLIT  = 4,
-    STATCOUNTERS_WRITE_RSP      = 5
+	const char *counter_name;
+	uint64_t (*counter_get)(void);
+} statcounter_names[] = {
+#define STATCOUNTER_ITEM(name, field, args) \
+	{ __XSTRING(name), &statcounters_read_##field },
+#include STATCOUNTERS_ARCH_INC
 };
 
 // helper functions
 
 #define ARCHNAME_BUFF_SZ 32
 
-static const char* getarchname(void)
+static const char *
+getarchname(void)
```

Comment:
Belongs in an MD header, and perhaps we should adopt the same mips64(-hybrid|-purecap) naming for MIPS too.

Reply:
Yeah probably, but I don't want to break old analysis scripts.